api/src/pcapi/alembic/CONTRIBUTING.md

```diff
@@ -124,6 +123,243 @@ Note : le `select 1` permet d'éviter une requête vide, ce que SQLAlchemy n'app
+3. `drop_constraint`
+
+#### Déplacement d'une colonne d'une table vers une autre table
+
+❌ Ne fonctionnera pas entre la migration *pre* et le déploiement du code N+1 :
+
+```python
+# pre/post deployment: pre
+def upgrade() -> None:
+    op.add_column("allocine_venue_provider", sa.Column("price", sa.Numeric(precision=10, scale=2)))
+
+# pre/post deployment: post
+def upgrade() -> None:
+    op.execute(backfill_data_from_a_table_to_another)
+    op.drop_column("allocine_venue_provider_price_rule", "price")
+
+# in the N+1 code, all the reads/writes go from allocine_venue_provider_price_rule.price to allocine_venue_provider.price
+```
+
+❓ Il existe une fenêtre de temps où la migration pre est effectuée mais le code N+1 n'est pas déployé. Dans cette fenêtre, il est possible d'écrire un `allocine_venue_provider` sans `price` parce que le code N n'a pas connaissance de cette colonne, ce qui entraîne un trou dans la raquette.
+
+✅ Deux déploiements doivent être réalisés :
+
+```python
+# first deployment
+# pre/post deployment: pre
+def upgrade() -> None:
+    op.add_column("allocine_venue_provider", sa.Column("price", sa.Numeric(precision=10, scale=2)))
+
+# in the N+1 code, allocine_venue_provider_price_rule.price is read and both columns are written
+
+# pre/post deployment: post
+def upgrade() -> None:
+    op.execute(backfill_data_from_a_table_to_another)
+```
+
+```python
+# in the N+2 code, all reads and writes are done in the new table
+
+# second deployment
+# pre/post deployment: post
+def upgrade() -> None:
+    op.drop_column("allocine_venue_provider_price_rule", "price")
+```
+
+Note : Déplacer une colonne de table ou renommer une colonne se comportent de la même manière.
+
+#### Ajout d'une contrainte
+
+❌ Risque de timeout si la table `cashflow` contient beaucoup (> 1 million) de lignes :
+
+```python
+op.create_foreign_key(
+    "stock_offererAddressId_fkey",
+    "stock",
+    "offerer_address",
+    ["offererAddressId"],
+    ["id"],
+)
+```
+
+❓ Ajouter une contrainte valide verrouille exclusivement la table, empêchant toute lecture et écriture en plus du risque de timeout.
+
+✅ La création et la validation de la contrainte doit être séparées dans une ou plusieurs migrations *post*. La création de la contrainte non-valide assure que la contrainte est respectée pour les nouvelles lignes mais pas les anciennes. La validation de la contrainte pour les lignes existantes se fait dans un second temps.
+
+```python
+op.create_foreign_key(
+    "stock_offererAddressId_fkey",
+    "stock",
+    "offerer_address",
+    ["offererAddressId"],
+    ["id"],
+    postgresql_not_valid=True,
+)
+
+# make sure that the constraint is enforced for all the existing table lines 
+
+op.execute("SET SESSION statement_timeout='300s'")  # or more if needed
+op.execute("""ALTER TABLE stock VALIDATE CONSTRAINT "stock_offererAddressId_fkey" """)
+op.execute(f"SET SESSION statement_timeout={settings.DATABASE_STATEMENT_TIMEOUT}")
+```
+
```

Comment:
Je suggère de mentionner `bin/alembic_add_not_null_constraint.py` qui génère les 4 "bonnes" migrations post-déploiement nécessaires à l'ajout d'une contrainte `not null` sur une table à fort volume ou qui est fréquemment modifiée.

Reply:
On pourrait aussi préciser qu'ajouter une contrainte de clé étrangère (sur une colonne existante ou lors de la création d'une nouvelle table, cas qu'on oublie régulièrement) demande un lock sur la table source **et sur la table cible**.

J'en ai parlé récemment sur Slack : 

> Petit rappel (car j'ai vu le même problème sur 2 pull requests cet après-midi) : quand on créé une table qui contient une clé étrangère, il faut penser qu'un lock exclusif doit être pris sur la table **cible** (la table référencée par la nouvelle table). Si cette table cible est fréquemment utilisée (`booking` , `user`, `offer`, etc. sont des bons candidats), la prise du lock peut échouer. Si cela arrive dans une migration **pré**-déploiement (ce qui est généralement le cas lors de la création d'une table), cela bloque la MEP.
>
> Il est préférable de créer, dans une migration **pré**-déploiement, la table **sans** la clé étrangère. Puis de créer la contrainte de clé étrangère dans une migration **post**-déploiement. Ainsi, si le lock ne peut pas être pris lors de la MEP, la MEP ne sera pas bloquée. Seules les migrations **post**-déploiement seront bloquées, ce qui ne devrait pas être un problème. On pourra facilement les passer à la main plus tard.

Et j'avais complété : 

> Le lock est posé sur la table cible (disons `user`). L'idée est de s'assurer qu'il n'y a pas une autre transaction en cours qui supprime la colonne cible (ou change son type ou autre chose) qui ferait que la clé étrangère définie dans la table source n'est plus possible. Certes, une fois le lock posé, il n'est pas "tenu" longtemps, puisque PostgreSQL n'a pas grand chose à faire. Mais la **prise** de ce lock peut être longue, par exemple s'il y a un fort volume de requêtes SQL et/ou des requêtes longues. Si c'est très long, ça peut même échouer, selon la valeur du "lock  timeout". 
>
> Hé, mais alors, on n'a qu'à augmenter ce "lock timeout". On le monte à 5 minutes, et comme ça on est à peu près certain que le lock finira par être pris. Fastoche, non ?
>
> Oui... mais ça va piquer. :)  Car toutes les requêtes SQL qui sont effectuées après la **demande** de lock (et avant sa libération) s'empilent sans être traitées par PostgreSQL. Donc si le lock met 3 minutes à être pris, ça veut dire qu'on va empiler toutes les requêtes (y compris de simples `select`) pendant 3 minutes. Si on a un statement timeout plus court que 3 minutes, toutes ces requêtes vont échouer (coucou les erreurs 500 en masse). Si on un statement timeout supérieur à 3 minutes, les requêtes ont des chances de finir par être exécutées, mais on aura explosé la charge sur Gunicorn parce que toutes les requêtes vont prendre plusieurs minutes. Donc certaines requêtes HTTP ne pourront pas être servies, donc des erreurs 503 en masse).